### PR TITLE
Add support for bundling dylib deps from CcInfo

### LINKED ISSUE #2472: Dylibs from CcInfo are not packaged with macos_application
A macos_application that has a cc_library as a dep does not correctly bundle referenced .dylibs.

For example, mylib.dylib should be placed in app.app/Contents/Frameworks but it isn't.
```
cc_import(
    name = "mylib",
    shared_library = "mylib.dylib",
)

cc_library(
    name = "libapp",
    deps = [":mylib"],
)

macos_application(
    name = "app",
    ...
    deps = [":libapp"],
)
```

### PR DESCRIPTION
**Pull request by gkoreman**
Fixes #2472

### CODE REVIEW

File: apple/internal/partials/cc_info_dylibs.bzl [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                        bundle_files.append(
+                            (processor.location.framework, None, depset([library.dynamic_library])),
+                        )
+
+    return struct(bundle_files = bundle_files)
+
+def cc_info_dylibs_partial(
+        *,
+        embedded_targets):
+    """Constructor for the CcInfo dylibs processing partial.
```
Comment (luispadron) [inline apple/internal/partials/cc_info_dylibs.bzl:47]: Could you add some info on what the partial does

File: test/starlark_tests/macos_application_tests.bzl [hunk 2]
```diff
@@ -445,6 +445,16 @@ def macos_application_test_suite(name):
         tags = [name],
     )
 
+    archive_contents_test(
```
Comment (luispadron) [inline test/starlark_tests/macos_application_tests.bzl:448]: Could we add another one of these for ios_application

Comment (gkoreman) [reply test/starlark_tests/macos_application_tests.bzl:448]: I don't think .dylibs are allowed to be packaged in the same way for ios_applications. If this is desired can any reviewers with experience in this area advise on how this should work?

File: apple/internal/partials/cc_info_dylibs.bzl (hunk 1 above)
Comment (gkoreman) [reply apple/internal/partials/cc_info_dylibs.bzl:47]: Added in ebf598b

File: test/starlark_tests/macos_application_tests.bzl (hunk 2 above)
Comment (luispadron) [reply test/starlark_tests/macos_application_tests.bzl:448]: Ah yeah I believe they'd need to be packaged as frameworks.

File: apple/internal/partials/cc_info_dylibs.bzl [hunk 3]
```diff
@@ -0,0 +1,70 @@
+# Copyright 2023 The Bazel Authors. All rights reserved.
```
Comment (brentleyjones) [inline apple/internal/partials/cc_info_dylibs.bzl:1]:
```suggestion
# Copyright 2024 The Bazel Authors. All rights reserved.
```

File: apple/internal/partials/cc_info_dylibs.bzl [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Partial implementation for gathering cc_info dylibs."""
+
+load(
+    "@bazel_skylib//lib:partial.bzl",
+    "partial",
+)
```
Comment (brentleyjones) [inline apple/internal/partials/cc_info_dylibs.bzl:20]:
```suggestion
load("@bazel_skylib//lib:partial.bzl", "partial")
```

File: apple/internal/partials/cc_info_dylibs.bzl [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+def _cc_info_dylibs_partial_impl(
+        *,
+        embedded_targets):
+    """Implementation for the CcInfo dylibs processing partial."""
+    bundle_files = []
+
+    for target in embedded_targets:
+        if CcInfo in target:
```
Comment (brentleyjones) [inline apple/internal/partials/cc_info_dylibs.bzl:33]: Can you change this to `if CcInfo not in target:` and have it `continue` to lower the nesting?

Comment (gkoreman) [reply apple/internal/partials/cc_info_dylibs.bzl]: Yep, done!

**Review by luispadron — APPROVED** [review]

File: test/starlark_tests/resources/ccinfo_dylibs/mylib/lib.hpp [hunk 6]
```diff
@@ -0,0 +1,3 @@
+#pragma once
+
+void MyLibFunc(const char *myString);
```
Comment (brentleyjones) [inline test/starlark_tests/resources/ccinfo_dylibs/mylib/lib.hpp:3]:
```suggestion
void MyLibFunc(const char *myString);

```

File: test/starlark_tests/resources/ccinfo_dylibs/mylib/lib.cpp [hunk 7]
```diff
@@ -0,0 +1,3 @@
+#include <iostream>
+
+void MyLibFunc(const char *myString) { std::cout << myString << "\n"; }
```
Comment (brentleyjones) [inline test/starlark_tests/resources/ccinfo_dylibs/mylib/lib.cpp:3]:
```suggestion
void MyLibFunc(const char *myString) { std::cout << myString << "\n"; }

```

**Review by brentleyjones — APPROVED** [review]

**Pull request merged**